ocaml/xapi/import.ml

```diff
@@ -2169,121 +2169,100 @@ let metadata_handler (req : Request.t) s _ =
-            (* CA-48768: Stream_vdi.recv_all only checks for task cancellation
-               						   every ten seconds, so we need to check again now. After this
-               						   point, we disable cancellation for this task. *)
-            TaskHelper.exn_if_cancelling ~__context ;
-            TaskHelper.set_not_cancellable ~__context ;
-            (* Pre-miami GA exports have a checksum table at the end of the export. Check the calculated checksums *)
-            (* against the table here. Nb. Rio GA-Miami B2 exports get their checksums checked twice! *)
-            ( if header.version.export_vsn < 2 then
-                let xml =
-                  Tar_unix.Archive.with_next_file s (fun s hdr -> read_xml hdr s)
-                in
-                let expected_checksums =
-                  xml |> Xmlrpc.of_string |> checksum_table_of_rpc
+          let vdis =
+            List.map
+              (fun x ->
+                let vdir =
+                  API.vDI_t_of_rpc (find_in_export x.id state.export)
                 in
-                if not (compare_checksums checksum_table expected_checksums)
-                then (
-                  error "Some data checksums were incorrect: VM may be corrupt" ;
-                  if not config.force then
-                    raise (IFailure Some_checksums_failed)
-                  else
-                    error
-                      "Ignoring incorrect checksums since 'force' flag was \
-                       supplied"
+                ( x.id
+                , lookup (Ref.of_string x.id) table
+                , vdir.API.vDI_virtual_size
                 )
-            ) ;
-            (* return vmrefs *)
-            Listext.List.setify
-              (List.map (fun (cls, id, r) -> Ref.of_string r) state.created_vms)
-          with e ->
-            Backtrace.is_important e ;
-            error "Caught exception during import: %s"
+                )
+              all_vdis
+          in
+          List.iter
+            (fun (extid, intid, size) ->
+              debug "Expecting to import VDI %s into %s (size=%Ld)" extid
+                (Ref.string_of intid) size
+              )
+            vdis ;
+          let checksum_table =
+            Stream_vdi.recv_all refresh_session s __context rpc session_id
+              header.version config.force vdis
+          in
+          (* CA-48768: Stream_vdi.recv_all only checks for task cancellation
+             						   every ten seconds, so we need to check again now. After this
+             						   point, we disable cancellation for this task. *)
+          TaskHelper.exn_if_cancelling ~__context ;
+          TaskHelper.set_not_cancellable ~__context ;
+          (* Pre-miami GA exports have a checksum table at the end of the export. Check the calculated checksums *)
+          (* against the table here. Nb. Rio GA-Miami B2 exports get their checksums checked twice! *)
+          ( if header.version.export_vsn < 2 then
```

Comment:
This is some deeply nested code. Could this be made flatter?

Reply:
I guess the next step might be to zap the code that handles pre-miami exports 🤣

Reply:
I think I might have some Rio exports lying around... perhaps I should quietly destroy them ;-)